File: packages/typescript/src/modules/paymentSystems.ts
/**
 * PaymentSystemsModule: fetch available payment methods for the merchant.
 */

import { mapReturnCodeToError } from '../errors/index.js';
import type { EcollectConfig } from '../config.js';
import { BASE_URLS } from '../config.js';
import type { GetPaymentSystemRequest, GetPaymentSystemResponse } from '../types/api.js';
import type { PaymentSystem, FiImage, FinancialInstitution } from '../types/index.js';
import type { HttpClient } from '../utils/http.js';
import type { SessionModule } from './session.js';

export class PaymentSystemsModule {
  private readonly config: EcollectConfig;
  private readonly http: HttpClient;
  private readonly session: SessionModule;

  constructor(config: EcollectConfig, http: HttpClient, session: SessionModule) {
    this.config = config;
    this.http = http;
    this.session = session;
  }

  /**
   * Retrieve all payment systems enabled for this merchant.
   */
  async getPaymentSystems(): Promise<PaymentSystem[]> {
    const url = `${BASE_URLS[this.config.environment]}/getPaymentSystem`;
    const sessionToken = await this.session.getActive();

    const body: GetPaymentSystemRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: sessionToken,
    };

    const res = await this.http.post<GetPaymentSystemRequest, GetPaymentSystemResponse>(url, body);

    if (res.ReturnCode === 'NO_RECORDS') return [];
    if (res.ReturnCode === 'FAIL_APIEXPIREDSESSION') {
      this.session.invalidate();
      body.SessionToken = await this.session.getActive();
      const retried = await this.http.post<GetPaymentSystemRequest, GetPaymentSystemResponse>(
        url,
        body,
      );
      if (retried.ReturnCode !== 'SUCCESS' && retried.ReturnCode !== 'NO_RECORDS') {
        throw mapReturnCodeToError(retried.ReturnCode);
      }
      return this._map(retried);
    }

    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return this._map(res);
  }

  private _map(res: GetPaymentSystemResponse): PaymentSystem[] {
    if (!res.PaymentSystemArray) return [];
    return res.PaymentSystemArray.map((ps) => ({
      paymentSystem: ps.PaymentSystem,
      brandImageUrl: ps.BrandImageUrl,
      fiImages: ps.FiImagesArray?.map(
        (fi): FiImage => ({
          fiCode: fi.FiCode,
          findKeys: fi.FindKeys,
          brandImageUrl: fi.BrandImageUrl,
        }),
      ),
      financialInstitutions: ps.FiArray?.map(
        (fi): FinancialInstitution => ({
          fiCode: fi.FiCode,
          fiName: fi.FiName,
        }),
      ),
    }));
  }
}

File: packages/python/ecollect/modules/payment_systems.py
"""Payment systems module (getPaymentSystem)."""
import logging
from typing import Any, Dict, List

from ecollect.config import EcollectConfig
from ecollect.exceptions import raise_for_return_code
from ecollect.utils.http import AsyncHttpClient, run_sync

logger = logging.getLogger(__name__)


class PaymentSystemsModule:
    """Query available payment methods for the merchant."""

    def __init__(
        self,
        config: EcollectConfig,
        http: AsyncHttpClient,
        session: "SessionModule",  # type: ignore[name-defined]
    ) -> None:
        self._config = config
        self._http = http
        self._session = session

    async def _post(self, endpoint: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        url = self._config.endpoint_url(endpoint)
        data = await self._http.post_with_retry(url, payload)
        return_code = data.get("ReturnCode", "")
        if return_code == "FAIL_APIEXPIREDSESSION":
            self._session.invalidate()
            payload["SessionToken"] = await self._session.get_active()
            data = await self._http.post_with_retry(url, payload)
        raise_for_return_code(data.get("ReturnCode", ""), str(data))
        return data

    async def get_payment_systems(self) -> List[Dict[str, Any]]:
        """Return the list of payment systems enabled for this merchant."""
        session_token = await self._session.get_active()
        payload = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
        }
        data = await self._post("getPaymentSystem", payload)
        return data.get("PaymentSystemArray", [])

    def get_payment_systems_sync(self) -> List[Dict[str, Any]]:
        return run_sync(self.get_payment_systems())
